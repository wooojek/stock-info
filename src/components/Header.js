import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { logout } from '../actions/auth';

export const Header = ({ logout }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/">
          <h1>Stock Info</h1>
        </Link>
        <Link className="header__element--link" to="/add">
          <h3>Add new company</h3>
        </Link>
        <button className="button button--link header__logout" onClick={logout}>Logout</button>
      </div>

    </div>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  logout: () => dispatch(logout()),
});

export default connect(undefined, mapDispatchToProps)(Header);