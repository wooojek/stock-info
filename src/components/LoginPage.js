import React from 'react';
import { connect } from 'react-redux';
import uuid from 'uuid/v1';
import { login } from '../actions/auth';

const uid = uuid();

export const LoginPage = ({ login }) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Stock info</h1>
      <p>Be up to date with your stocks!</p>
      <button className="button" onClick={login}>Login</button>
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  login: () => dispatch(login(uid)),
});

export default connect(undefined, mapDispatchToProps)(LoginPage);