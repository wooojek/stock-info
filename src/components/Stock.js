import React from 'react';
import numeral from 'numeral';

const { ALPHAVANTAGE_API_KEY } = process.env;

export class Stock extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name: props.name,
      symbol: props.symbol,
      market: props.market,
      domain: props.domain || '',
      logo: props.logo || 'https://via.placeholder.com/64',
    }
  }

  getImage = () => {
    const query = this.state.name.split(' ')[0];
    return fetch(`https://autocomplete.clearbit.com/v1/companies/suggest?query=${query}`)
      .then((response) => {
        if (!response.ok) {
          throw Error(response.statusText);
        }
        return response.json();
      })
      .then(([{ domain, logo }]) => {
        return {
          domain,
          logo,
        };
      })
      .catch((e) => console.log(e));
  }

  getStockInfo = () => {
    return fetch(`https://www.alphavantage.co/query?function=GLOBAL_QUOTE&symbol=${this.state.symbol}&apikey=${ALPHAVANTAGE_API_KEY}`)
        .then((response) => {
          if (!response.ok) {
            throw Error(response.statusText);
          }
          return response.json();
        })
        .then((data) => {
          if (data['Note']) {
            throw Error(data['Note']);
          }
          const stockData = data['Global Quote'];
          return {
            price: stockData['05. price'],
            closed: stockData['07. latest trading day'],
            change: stockData['09. change'],
            changePercent: stockData['10. change percent'],
          };
        })
        .catch((e) => console.log(e));
  }

  async componentDidMount() {
    try {
      const { domain, logo } = await this.getImage();
      const details = await this.getStockInfo();

      this.setState(() => ({
        details,
        domain,
        logo,
      }));
    } catch (e) {
      console.log(e);
    }
  }

  render() {
    const details = this.state.details;
    const market = this.state.market;
    return (
      <div className="list-item">
        <div className="stock-container">
          <div
            className="stock-img"
            style={{ backgroundImage: `url(${this.state.logo})` }}
          >
          </div>
          <div className="stock-info">
            <div className="stock-info__names">
              <span className="bold">{this.state.name}</span>
              <span>{this.state.symbol}</span>
              {this.state.domain ? (<span>{this.state.domain}</span>) : false}
            </div>
            {market ? (
              <div className="stock-info__market">
                <span>{market.region}</span>
                <span>{market.marketOpen} - {market.marketClose}</span>
                <span>{market.timezone}</span>
              </div>
            ) : (<div></div>)}
            {details ? (
              <div className="stock-info__details">
                <span className="bold">{numeral(details.price).format('$0,0.00')}</span>
                <span className={details.change >= 0 ? 'stock--up' : 'stock--down'}>
                  {numeral(details.change).format('0.00')} ({numeral(details.changePercent.slice(0, -2)).format('0.00')}%)
                </span>
                <span>Closed: {details.closed}</span>
              </div>
            ) : (<div></div>)}

          </div>
        </div>
      </div>
    )
  }
}

export default Stock;