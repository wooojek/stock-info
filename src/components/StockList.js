import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

export const StockList = ({ stocks }) => (
  <Fragment>
    <h2>Companies</h2>
    <div className="content-container">
      {stocks.length > 0 ?
        (
          <ul className="list-body">
            {stocks.map((stock, index) => <li className="list-item" key={index}>{stock.name}</li>)}
          </ul>
        ) : (
          <div>
            <span>There are no companies yet.</span>
            <Link to="/add">
              <p>Track your first company.</p>
            </Link>
          </div>
        )
      }
    </div>
  </Fragment>
);

const mapStateToProps = (state) => ({
  stocks: state.stocks,
});

export default connect(mapStateToProps)(StockList);