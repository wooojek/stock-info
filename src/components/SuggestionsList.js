import React from 'react';

export const SuggestionsList = ({ suggestions, onSelect }) => {

  const onClick = (sugestion) => {
    onSelect({
      symbol: sugestion['1. symbol'],
      name: sugestion['2. name'],
    });
  }

  return (
    <div className="suggestions-container">
      <ul className='list-body suggestions-cover'>
        {suggestions.map((sugestion, index) => (
          <li
            key={index}
            className='list-item'
            onClick={() => onClick(sugestion)}
          >
            {sugestion['1. symbol']} - {sugestion['2. name']}
          </li>
        ))}
      </ul>
    </div>
  );
};

export default SuggestionsList;