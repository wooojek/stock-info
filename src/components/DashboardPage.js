import React from 'react';
import StockList from './StockList';

export const DashboardPage = () => (
  <div className='content-container'>
    <StockList />
  </div>
);

export default DashboardPage;