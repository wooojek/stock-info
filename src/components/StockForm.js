import React from 'react';
import SuggestionsList from './SuggestionsList';

const { ALPHAVANTAGE_API_KEY } = process.env;

export default class StockForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name: props.stock ? props.stock.name : '',
      symbol: props.stock ? props.stock.symbol : '',
      suggestions: [],
      error: '',
    }
  }

  onNameChange = (e) => {
    const name = e.target.value.toUpperCase();
    this.setState(() => ({ name }), () => {
      if (this.state.name && this.state.name.length > 1) {
        if (this.state.name.length % 2 === 0) {
          this.getSuggestions();
        }
      }
    });
  }

  onNameSelect = ({ name, symbol }) => {
    const suggestions = [];
    this.setState(() => ({ name, suggestions, symbol }));
  }

  getSuggestions = () => {
    fetch(`https://www.alphavantage.co/query?function=SYMBOL_SEARCH&keywords=${this.state.name}&apikey=${ALPHAVANTAGE_API_KEY}`)
      .then((response) => response.json())
      .then(({ bestMatches }) => {
        this.setState(() => ({
          suggestions: bestMatches.slice(0, 4),
        }));
      });
  }

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.name) {
      this.setState(() => ({ errorState: 'Please provide name.' }));
    } else {
      this.setState(() => ({ errorState: '' }));
      this.props.onSubmit({
        name: this.state.name,
        symbol: this.state.symbol,
      });
    }
  }

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.errorState && (<p className="form__error">{this.state.errorState}</p>)}
        <input
          type="text"
          className="text-input"
          placeholder="Company symbol"
          autoFocus
          value={this.state.name}
          onChange={this.onNameChange}
        />
        <SuggestionsList suggestions={this.state.suggestions} onSelect={this.onNameSelect} />
        <div>
          <button className="button">Track</button>
        </div>
      </form>
    )
  }
};