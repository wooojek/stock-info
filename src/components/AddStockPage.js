import React from 'react';
import { connect } from 'react-redux';
import StockForm from './StockForm';
import { startAddStock } from '../actions/stocks';

export class AddStockPage extends React.Component {
  onSubmit = (stock) => {
    this.props.startAddStock(stock);
    this.props.history.push('/')
  }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2 className="page-header__title">Add Stock</h2>
          </div>
        </div>
        <div className="content-container">
          <StockForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    )
  }
};

const mapDispatchToProps = (dispatch) => ({
  startAddStock: (stock) => dispatch(startAddStock(stock)),
});

export default connect(undefined, mapDispatchToProps)(AddStockPage);