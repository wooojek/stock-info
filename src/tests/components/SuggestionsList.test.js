import React from 'react';
import { shallow } from 'enzyme';
import { SuggestionsList } from '../../components/SuggestionsList';
import suggestions from '../fixtures/suggestions';

test('should render SuggestionsList correctly', () => {
  const wrapper = shallow(<SuggestionsList suggestions={suggestions} onSelect={() => {}} />);
  expect(wrapper).toMatchSnapshot();
});

test('should call onSelect on button click', () => {
  const onSelect = jest.fn();
  const wrapper = shallow(<SuggestionsList suggestions={suggestions} onSelect={onSelect} />);
  wrapper.find('li').at(0).simulate('click');
  expect(onSelect).toHaveBeenCalled();
});