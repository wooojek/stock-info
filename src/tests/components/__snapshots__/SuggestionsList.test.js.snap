// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SuggestionsList correctly 1`] = `
<div
  className="suggestions-container"
>
  <ul
    className="list-body suggestions-cover"
  >
    <li
      className="list-item"
      key="0"
      onClick={[Function]}
    >
      AAPL
       - 
      Apple Inc.
    </li>
    <li
      className="list-item"
      key="1"
      onClick={[Function]}
    >
      AAPL.MEX
       - 
      Apple Inc.
    </li>
    <li
      className="list-item"
      key="2"
      onClick={[Function]}
    >
      AAPL34.SAO
       - 
      Apple Inc.
    </li>
    <li
      className="list-item"
      key="3"
      onClick={[Function]}
    >
      AAPL.ARG
       - 
      Apple Inc.
    </li>
  </ul>
</div>
`;
